File: UI Code/src/tutor/bookSearch.ts
import type { TextBook } from "./book";
import { TEACHING_KINDS } from "./outline";

export type SearchHit = { pdfPage: number; score: number };

/** Offline keyword search (BM25) over the page packages, so the teacher (and the search screen)
 *  can find the right pages anywhere in the book. No model, no network.
 *  Port of app/lib/tutor/book_search.dart. */
export class BookSearch {
  private docs = new Map<number, Map<string, number>>(); // pdf page -> term frequencies
  private topics = new Map<number, Set<string>>(); // pdf page -> words of its section, terms, chapter
  private lengths = new Map<number, number>();
  private docFreq = new Map<string, number>();
  private avgLength = 1;

  constructor(readonly book: TextBook) {
    for (let page = 1; page <= book.pageCount; page++) {
      const pkg = book.pkg(page);
      if (!pkg || !TEACHING_KINDS.has(pkg.pageKind)) continue;
      const tf = new Map<string, number>();
      const add = (text: unknown, weight = 1) => {
        if (typeof text !== "string") return;
        for (const t of tokenize(text)) tf.set(t, (tf.get(t) ?? 0) + weight);
      };
      const chapterTitle = book.chapter(pkg.chapter)?.titleFa;
      add(pkg.sectionFa, 3);
      add(chapterTitle);
      for (const term of pkg.terms) {
        add(term.fa, 2);
        add(term.en, 2);
      }
      pkg.keyPoints.forEach((k) => add(k));
      add(pkg.summaryFa);
      for (const f of pkg.formulas) add(f.meaning_fa);
      for (const f of pkg.figures) add(f.description_fa);
      for (const e of [...pkg.workedExamples, ...pkg.exercises]) add(e.problem_fa);
      if (!tf.size) continue;
      this.docs.set(page, tf);
      this.topics.set(page, new Set([
        ...tokenize(pkg.sectionFa ?? ""),
        ...tokenize(chapterTitle ?? ""),
        ...pkg.terms.flatMap((t) => [...tokenize(t.fa ?? ""), ...tokenize(t.en ?? "")]),
      ]));
      let len = 0;
      tf.forEach((v) => (len += v));
      this.lengths.set(page, len);
      for (const t of tf.keys()) this.docFreq.set(t, (this.docFreq.get(t) ?? 0) + 1);
    }
    if (this.lengths.size) {
      let sum = 0;
      this.lengths.forEach((v) => (sum += v));
      this.avgLength = sum / this.lengths.size;
    }
  }

  /** The search terms that occur on a page (empty for pages that aren't indexed). */
  termsOf(pdfPage: number): Set<string> {
    return new Set(this.docs.get(pdfPage)?.keys() ?? []);
  }

  /** Words naming what a page is about: its section title, its terms and its chapter title. */
  topicTermsOf(pdfPage: number): Set<string> {
    return this.topics.get(pdfPage) ?? new Set();
  }

  /** Whether a term occurs anywhere in the book. */
  knows(term: string): boolean {
    return this.docFreq.has(term);
  }

  /** Every page sharing a term with the query, best first. */
  search(query: string): SearchHit[] {
    const terms = new Set(tokenize(query));
    const n = this.docs.size;
    const k1 = 1.2, b = 0.75;
    const scores = new Map<number, number>();
    for (const t of terms) {
      const df = this.docFreq.get(t);
      if (df === undefined) continue;
      const idf = Math.log(1 + (n - df + 0.5) / (df + 0.5));
      this.docs.forEach((tf, page) => {
        const f = tf.get(t);
        if (f === undefined) return;
        const norm = (f * (k1 + 1)) / (f + k1 * (1 - b + (b * this.lengths.get(page)!) / this.avgLength));
        scores.set(page, (scores.get(page) ?? 0) + idf * norm);
      });
    }
    return [...scores.entries()].map(([pdfPage, score]) => ({ pdfPage, score })).sort((a, b) => b.score - a.score);
  }
}

const LATEX_COMMAND = /\\[a-zA-Z]+/g;
const SEPARATOR = /[^a-z0-9ؠ-يٮ-ۓ]+/;

/** Folds Arabic/Persian letter variants, digits and diacritics so the same word always looks the
 *  same; ZWNJ becomes a space so plural/verb affixes split off (انتیگرال‌ها → انتیگرال ها). */
export function normalizeForSearch(s: string): string {
  let out = "";
  for (const ch of s.toLowerCase().replace(LATEX_COMMAND, " ")) {
    const r = ch.codePointAt(0)!;
    if (r === 0x064a || r === 0x0649) out += "ی"; // ي ى → ی
    else if (r === 0x0643) out += "ک"; // ك → ک
    else if (r === 0x0629 || r === 0x06c0) out += "ه"; // ة ۀ → ه
    else if (r === 0x0623 || r === 0x0625 || r === 0x0671) out += "ا"; // أ إ ٱ → ا
    else if ((r >= 0x064b && r <= 0x065f) || r === 0x0670 || r === 0x0640) continue; // diacritics, tatweel
    else if (r >= 0x06f0 && r <= 0x06f9) out += String.fromCharCode(0x30 + r - 0x06f0);
    else if (r >= 0x0660 && r <= 0x0669) out += String.fromCharCode(0x30 + r - 0x0660);
    else if (r === 0x200c) out += " ";
    else out += ch;
  }
  return out;
}

/** Search tokens: normalized words minus stopwords and bare numbers, plural suffixes stripped. */
export function tokenize(text: string): string[] {
  const tokens: string[] = [];
  for (const raw of normalizeForSearch(text).split(SEPARATOR)) {
    if (raw.length < 2 || STOPWORDS.has(raw) || /^[0-9]+$/.test(raw)) continue;
    const t = stem(raw);
    if (!STOPWORDS.has(t)) tokens.push(t);
  }
  return tokens;
}

function stem(t: string): string {
  for (const suffix of ["های", "ها", "ات"]) {
    if (t.length > suffix.length + 2 && t.endsWith(suffix)) return t.slice(0, t.length - suffix.length);
  }
  return t;
}

// Function words, question words and words that appear on nearly every page (in normalized form).
const STOPWORDS = new Set([
  "و", "در", "به", "از", "که", "این", "ان", "آن", "را", "با", "است", "اند", "هست", "هستند", "بود", "برای",
  "یک", "یا", "تا", "بر", "هم", "نیز", "می", "شود", "شده", "شد", "کنید", "کند", "کنیم", "کنم", "کرد",
  "کردن", "کنند", "باید", "اگر", "هر", "چه", "چی", "چطور", "چگونه", "چرا", "کجا", "کدام", "آیا", "ایا",
  "لطفا", "بدهید", "بگویید", "بده", "بگو", "توضیح", "دهید", "دهم", "ساده", "صفحه", "صفحات", "کتاب",
  "درس", "مورد", "باره", "درباره", "ما", "شما", "من", "او", "ها", "های", "ای", "یعنی", "چیست",
  "چیه", "چند", "دارد", "دارند", "داریم", "دیگر", "همین", "اینجا", "آنجا", "میشود", "بفهمم", "نمیفهمم",
  "فهمیدم", "حل", "سوال", "سوالی", "مثال", "تمرین", "فصل", "اول", "دوم", "سوم", "چهارم", "پنجم", "ششم",
  "هفتم", "هشتم", "بپرسید", "بیشتر", "خوب", "بخش", "قسمت", "مطلب", "چیزی", "وقتی", "چون", "پس", "اما",
  "ولی", "نه", "بلی", "باشد", "میتوانم", "میتوان", "توانم", "نشان", "میدهد", "دهد", "داده", "خوانم",
  "بخوانم", "میگوید", "گوید", "چیزهایی", "موضوع", "the", "of", "and", "is", "to", "in", "what",
  // Common phrasing in questions ("به زبان ساده", "به دست آمد", "شکل پایین صفحه", "قدم به قدم").
  "زبان", "دست", "آمد", "آمده", "آید", "آورد", "آورده", "آوردن", "کار", "برد", "بردن", "برده", "گرفت",
  "گرفته", "گیرد", "شوند", "بشود", "حساب", "شکل", "پایین", "بالا", "نوشته", "فهم", "کمک", "راه", "روش",
  "مرحله", "قدم", "بعدی", "قبلی", "همه", "کل",
]);

File: app/lib/tutor/book_search.dart
import 'dart:math' as math;

import '../content/book.dart';

class SearchHit {
  const SearchHit(this.pdfPage, this.score);

  final int pdfPage;
  final double score;
}

/// Offline keyword search (BM25) over the page packages, so the teacher can find the right pages
/// anywhere in the book. Needs no model or network, and runs the same way on Android.
/// Needs [Book.preloadPackages] to have run.
class BookSearch {
  BookSearch(this.book) {
    _build();
  }

  final Book book;
  final _docs = <int, Map<String, int>>{}; // pdf page -> term frequencies
  final _topics = <int, Set<String>>{}; // pdf page -> words of its section title, terms and chapter
  final _lengths = <int, int>{};
  final _docFreq = <String, int>{};
  double _avgLength = 1;

  static const _teachingKinds = {'content', 'exercises', 'review'};

  void _build() {
    for (var page = 1; page <= book.pageCount; page++) {
      final pkg = book.cachedPackage(page);
      if (pkg == null || !_teachingKinds.contains(pkg.json['page_kind'])) continue;
      final tf = <String, int>{};
      void add(Object? text, [int weight = 1]) {
        if (text is! String) return;
        for (final t in tokenize(text)) {
          tf[t] = (tf[t] ?? 0) + weight;
        }
      }

      add(pkg.sectionFa, 3);
      add(book.chapter(pkg.chapter)?.titleFa);
      for (final term in (pkg.json['terms'] as List? ?? const [])) {
        add((term as Map)['fa'], 2);
        add(term['en'], 2);
      }
      pkg.keyPoints.forEach(add);
      add(pkg.summaryFa);
      for (final f in pkg.formulas) {
        add(f['meaning_fa']);
      }
      for (final f in pkg.figures) {
        add(f['description_fa']);
      }
      for (final e in [...pkg.workedExamples, ...pkg.exercises]) {
        add(e['problem_fa']);
      }
      if (tf.isEmpty) continue;
      _docs[page] = tf;
      _topics[page] = {
        ...tokenize(pkg.sectionFa ?? ''),
        ...tokenize(book.chapter(pkg.chapter)?.titleFa ?? ''),
        for (final term in (pkg.json['terms'] as List? ?? const [])) ...[
          ...tokenize((term as Map)['fa'] as String? ?? ''),
          ...tokenize(term['en'] as String? ?? ''),
        ],
      };
      _lengths[page] = tf.values.fold(0, (a, b) => a + b);
      for (final t in tf.keys) {
        _docFreq[t] = (_docFreq[t] ?? 0) + 1;
      }
    }
    if (_lengths.isNotEmpty) {
      _avgLength = _lengths.values.reduce((a, b) => a + b) / _lengths.length;
    }
  }

  /// The search terms that occur on [pdfPage] (empty for pages that aren't indexed).
  Set<String> termsOf(int pdfPage) => _docs[pdfPage]?.keys.toSet() ?? const {};

  /// Words naming what [pdfPage] is about: its section title, its terms and its chapter title.
  Set<String> topicTermsOf(int pdfPage) => _topics[pdfPage] ?? const {};

  /// Whether [term] occurs anywhere in the book.
  bool knows(String term) => _docFreq.containsKey(term);

  /// Every page sharing a term with [query], best first.
  List<SearchHit> search(String query) {
    final terms = tokenize(query).toSet();
    final n = _docs.length;
    const k1 = 1.2, b = 0.75;
    final scores = <int, double>{};
    for (final t in terms) {
      final df = _docFreq[t];
      if (df == null) continue;
      final idf = math.log(1 + (n - df + 0.5) / (df + 0.5));
      _docs.forEach((page, tf) {
        final f = tf[t];
        if (f == null) return;
        final norm = f * (k1 + 1) / (f + k1 * (1 - b + b * _lengths[page]! / _avgLength));
        scores[page] = (scores[page] ?? 0) + idf * norm;
      });
    }
    return [for (final e in scores.entries) SearchHit(e.key, e.value)]
      ..sort((a, b) => b.score.compareTo(a.score));
  }
}

final _latexCommand = RegExp(r'\\[a-zA-Z]+');
final _separator = RegExp(r'[^a-z0-9ؠ-يٮ-ۓ]+');

/// Folds Arabic/Persian letter variants, digits and diacritics so that the same word always
/// looks the same; ZWNJ becomes a space so plural/verb affixes split off (انتیگرال‌ها → انتیگرال ها).
String normalizeForSearch(String s) {
  final out = StringBuffer();
  for (final r in s.toLowerCase().replaceAll(_latexCommand, ' ').runes) {
    if (r == 0x064A || r == 0x0649) {
      out.writeCharCode(0x06CC); // ي ى → ی
    } else if (r == 0x0643) {
      out.writeCharCode(0x06A9); // ك → ک
    } else if (r == 0x0629 || r == 0x06C0) {
      out.writeCharCode(0x0647); // ة ۀ → ه
    } else if (r == 0x0623 || r == 0x0625 || r == 0x0671) {
      out.writeCharCode(0x0627); // أ إ ٱ → ا
    } else if ((r >= 0x064B && r <= 0x065F) || r == 0x0670 || r == 0x0640) {
      // diacritics and tatweel: drop
    } else if (r >= 0x06F0 && r <= 0x06F9) {
      out.writeCharCode(0x30 + r - 0x06F0);
    } else if (r >= 0x0660 && r <= 0x0669) {
      out.writeCharCode(0x30 + r - 0x0660);
    } else if (r == 0x200C) {
      out.write(' ');
    } else {
      out.writeCharCode(r);
    }
  }
  return out.toString();
}

/// Search tokens: normalized words minus stopwords and bare numbers, with plural suffixes stripped.
List<String> tokenize(String text) {
  final tokens = <String>[];
  for (final raw in normalizeForSearch(text).split(_separator)) {
    if (raw.length < 2 || _stopwords.contains(raw) || RegExp(r'^[0-9]+$').hasMatch(raw)) continue;
    final t = _stem(raw);
    if (!_stopwords.contains(t)) tokens.add(t);
  }
  return tokens;
}

String _stem(String t) {
  for (final suffix in const ['های', 'ها', 'ات']) {
    if (t.length > suffix.length + 2 && t.endsWith(suffix)) return t.substring(0, t.length - suffix.length);
  }
  return t;
}

// Function words, question words and words that appear on nearly every page (in normalized form).
const _stopwords = {
  'و', 'در', 'به', 'از', 'که', 'این', 'ان', 'آن', 'را', 'با', 'است', 'اند', 'هست', 'هستند', 'بود', 'برای',
  'یک', 'یا', 'تا', 'بر', 'هم', 'نیز', 'می', 'شود', 'شده', 'شد', 'کنید', 'کند', 'کنیم', 'کنم', 'کرد',
  'کردن', 'کنند', 'باید', 'اگر', 'هر', 'چه', 'چی', 'چطور', 'چگونه', 'چرا', 'کجا', 'کدام', 'آیا', 'ایا',
  'لطفا', 'بدهید', 'بگویید', 'بده', 'بگو', 'توضیح', 'دهید', 'دهم', 'ساده', 'صفحه', 'صفحات', 'کتاب',
  'درس', 'مورد', 'باره', 'بارهٔ', 'درباره', 'ما', 'شما', 'من', 'او', 'ها', 'های', 'ای', 'یعنی', 'چیست',
  'چیه', 'چند', 'دارد', 'دارند', 'داریم', 'دیگر', 'همین', 'اینجا', 'آنجا', 'میشود', 'بفهمم', 'نمیفهمم',
  'فهمیدم', 'حل', 'سوال', 'سوالی', 'مثال', 'تمرین', 'فصل', 'اول', 'دوم', 'سوم', 'چهارم', 'پنجم', 'ششم',
  'هفتم', 'هشتم', 'بپرسید', 'بیشتر', 'خوب', 'بخش', 'قسمت', 'مطلب', 'چیزی', 'وقتی', 'چون', 'پس', 'اما',
  'ولی', 'نه', 'بلی', 'باشد', 'میتوانم', 'میتوان', 'توانم', 'نشان', 'میدهد', 'دهد', 'داده', 'خوانم',
  'بخوانم', 'میگوید', 'گوید', 'چیزهایی', 'موضوع', 'the', 'of', 'and', 'is', 'to', 'in', 'what',
  // Common phrasing in questions ("به زبان ساده", "به دست آمد", "شکل پایین صفحه", "قدم به قدم").
  'زبان', 'دست', 'آمد', 'آمده', 'آید', 'آورد', 'آورده', 'آوردن', 'کار', 'برد', 'بردن', 'برده', 'گرفت',
  'گرفته', 'گیرد', 'شوند', 'بشود', 'حساب', 'شکل', 'پایین', 'بالا', 'نوشته', 'فهم', 'کمک', 'راه', 'روش',
  'مرحله', 'قدم', 'بعدی', 'قبلی', 'همه', 'کل',
};
